packages/opencensus-core/src/tags/tag-map.ts

```diff
@@ -0,0 +1,75 @@
+/**
+ * Copyright 2019, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 the "License";
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {TagKey, TagValue} from './types';
+import {isValidTagKey, isValidTagValue} from './validation';
+
+/** TagMap is maps of TagKey -> TagValue */
+export class TagMap {
+  // A map mapping TagKey to to its respective TagValue.
+  private readonly registeredTags: Map<TagKey, TagValue> = new Map();
+
+  constructor() {}
+
+  /**
+   * Inserts a key and value in the map if the map does not already contain the
+   * key.
+   * @param tagKey The tag key.
+   * @param tagValue The tag value.
+   */
+  insert(tagKey: TagKey, tagValue: TagValue): void {
+    if (!isValidTagKey(tagKey)) {
+      throw Error(`Invalid TagKey name: ${tagKey.name}`);
+    }
+
+    if (!isValidTagValue(tagValue)) {
+      throw Error(`Invalid TagValue: ${tagValue.value}`);
+    }
+
+    if (!this.registeredTags.has(tagKey)) {
+      this.registeredTags.set(tagKey, tagValue);
+    }
+  }
+
+  /**
+   * Deletes a tag from the map if the key is in the map.
+   * @param tagKey The tag key.
+   */
+  delete(tagKey: TagKey): void {
+    if (this.registeredTags.has(tagKey)) {
+      this.registeredTags.delete(tagKey);
+    }
+  }
+
+  /**
+   * Updates the map by updating the value of a key.
+   * @param tagKey A tag key to be updated.
+   * @param tagValue The value to update the key to in the map.
+   */
+  update(tagKey: TagKey, tagValue: TagValue): void {
```

Comment:
Why have separate `insert` and `update` functions and not just a `set` function that does insert and update?

That would enable combining both functions and you also would not need the `if (this.registeredTags.has(tagKey)) {` check.

Reply:
+1 to this. In Java we use this approach in [`TagContextBuilder.put()`](https://github.com/census-instrumentation/opencensus-java/blob/master/impl_core/src/main/java/io/opencensus/implcore/tags/TagContextBuilderImpl.java#L40).

Reply:
Done, thanks for the reference.